### PR TITLE
Style hierarchy

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @brson

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -93,3 +93,4 @@ tmp.*.rs
 version.md
 version.ml
 version.texi
+src/librustdoc/html/styles.rs
```
Comment (apasel422) [inline .gitignore]: Could you add the newline back to the end of this file?

**Comment by brson** [comment]
From our conversation on IRC I thought this was going to be a refactoring of the CSS to make it easy to customize, without explicit support for customization.

**Comment by brson** [comment]
Previous PR: https://github.com/rust-lang/rust/pull/29703

**Comment by GuillaumeGomez** [comment]
@brson: The CSS part to customize the doc interface (its colors to be even more precise) has been splitted out from the big file. So it doesn't change anything for people who don't care about adding new theme and give access to the functionality for others.

File: .gitignore (hunk 1 above)
Comment (GuillaumeGomez) [inline .gitignore]: Done.

**Comment by brson** [comment]
I still don't feel confident about this, and am going to wait to talk to the tools team again.

**Comment by GuillaumeGomez** [comment]
@brson: No problem, we're not in a hurry.

**Comment by brson** [comment]
@GuillaumeGomez can you remind me the difference between this PR and the previous?

**Comment by GuillaumeGomez** [comment]
This one just adds the possibility to add custom theme. The JS and the CSS are ready, but nothing change for the actual render (since there is only the "main" theme).

**Comment by brson** [comment]
@GuillaumeGomez thanks again for your patience on this issue. 

Even though I personally like dark themes, and want to be able to
accomodate your use cases, I am still uneasy about this direction, and
after talking it over with several other team members, I am going to
close this PR.

It's a difficult decision, and I'll admit up front that the rationale for
not accepting it is soft, but it is as follows.

First, this opens the door to proliferation of visual design across
the ecosystem.

Second, as with other attempts to overhaul the visual design of
specific aspects of rust's web presense, we are reluctant to add
features or make major changes, even with limited scope, largely
because we don't have any firm direction when it comes to the overall
visual design of our web properties - we don't have a 'vision' in this
area. We recognize this is a problem for the project and want to fix
it, but are not prepared to do so now.

On the technical side, while not a large patch, this adds code
generation to the rustdoc build system, a complicated solution for a
minor feature that upstream won't even use.

Complexity accumulates one feature at a time; this is a feature we do
not want right now, and accepting it even though its optional
increases the overall complexity of the (already messy) rustdoc
architecture. Reluctantly accepting patches we don't want to maintain
is not something we can do in general, and not something we will do in
this case. Doing so has bit us many times in the past, resulting in
poor design that had to be painfully reversed later. Sometimes we have
to draw the line and say 'no'.

I suggest that those that want to customize the rustdoc theme for now
should use browser-based css customization. We will happily consider
changes to CSS to accomodate such client-side customization.

**Pull request closed without merging**